### PR TITLE
PR-Template: Updated the PR template with new checkboxes

### PR DESCRIPTION
**Pull request by minosgalanakis**
## Description

Updates the development branch's PR checkboxes as discussed in [PR #9008](https://github.com/Mbed-TLS/mbedtls/pull/9008). Will need to apply for 3.6 and 2.28 braches as well before merging.

### CODE REVIEW

**Comment by gilles-peskine-arm** [comment]
I've set this PR (in its current state) as the default branch on my fork so you can see the template in action at https://github.com/gilles-peskine-arm/mbedtls/compare/development...gilles-peskine-arm:mbedtls:pr_9080

> Will need to apply for 3.6 and 2.28 braches as well before merging.

No: the template is only ever read from the default branch. We can't have different templates based on a pull request's target branch.

File: .github/pull_request_template.md
Comment (yanesca) [inline .github/pull_request_template.md:9]:
```suggestion
Please add/remove strikethrough ~~ as appropriate, and add any relevant link/s to the end of the line.
```

**Review by yanesca — CHANGES_REQUESTED** [review]
Just a minor typo but other than that it looks good

**Review by yanesca — DISMISSED** [review]
LGTM

**Comment by mpg** [comment]
Note: after merging this, we'll want to merge development to development-restricted (the default branch of the restricted repo) so that the new template is used there as well.

File: .github/pull_request_template.md
Comment (mpg) [inline .github/pull_request_template.md:14]: Markdown: `**2.28 PR**`

File: .github/pull_request_template.md
Comment (mpg) [inline .github/pull_request_template.md:9]: Not clear to me: if I want to add a justification, where should I add it? Before of after the link?

File: .github/pull_request_template.md
Comment (mpg) [inline .github/pull_request_template.md:12]: Question (does not have to result in a change in this PR, I just want to make sure I understand): for the development PR, will we add a link to self here?

**Review by mpg — CHANGES_REQUESTED** [review]
Looks pretty good to me other than a markdown typo.

Just one thing: is the template supposed to be self-sufficient or do we expect to have clarifying documentation elsewhere in addition (like, added to one of our guides about the review process)? For example, about the fact the checking the boxes should be done by the gatekeeper, not the PR author or reviewers - but the PR author should provide justifications that reviewers should review.

I don't have a strong opinion about where it should be documented, but I really think this kind of thing should be put in writing somewhere in (other than comments on #9008 I mean).

File: .github/pull_request_template.md
Comment (gilles-peskine-arm) [reply .github/pull_request_template.md:12]: Regardless, I expect the tooling to work no matter where the link is on the line, and no matter what the source syntax is (`#nnnn` or a bare link or a Markdown link).

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
I've done a first round of review. I would like several changes to make the template focus on useful information and more likely to be followed.

File: .github/pull_request_template.md
Comment (minosgalanakis) [reply .github/pull_request_template.md:9]: Originally the [proposal was](https://github.com/Mbed-TLS/mbedtls/pull/9008#issuecomment-2084739951) to have a comments section after the special code.

That way it would be easier to parse as a block of text without affecting the template.

e.g. 

~~~~~~
- [ ] **dev PR** provided, ~~not required~~ (Justification/Comments). Link --> #XYZ
~~~~~~

But I removed it since the comment's suggested that we should not increase the text of the checkboxes.

I am happy either way.

File: .github/pull_request_template.md
Comment (minosgalanakis) [reply .github/pull_request_template.md:12]: @mpg Ideally that should only be needed in backport-prs. So for a development PR if we include the comments that link could be

~~~~~~
- [ x ] **dev PR** ~~provided~~, not required. This is the dev PR.  Link --> #XYZ
~~~~~~~

But as @gilles-peskine-arm the tooling should work even if someone does that

~~~~~~
- [ x ] **dev PR** provided, ~~not required~~.  Link --> #SELF_NO
~~~~~~~

File: .github/pull_request_template.md
Comment (minosgalanakis) [reply .github/pull_request_template.md:12]: The purpose of the `Link --> #XYZ` was to make clear to the user what its meant to do. I think `link: #` would be a good compromise, and after people are confident with it we cna remove the link part. 

The tool does not care since it will be looking for markdown http or #{0-9} type of linkage in the PR's body anyway. 

I can change the order of items like not `required, provided`, and `note required, covered provided` for changelog entries, but I do not think it should matter that much,

File: .github/pull_request_template.md
Comment (minosgalanakis) [reply .github/pull_request_template.md:11]: I am with you on that. In unique cases that it is covered by other PR's it can be manually reviewed.

Comment (minosgalanakis) [reply .github/pull_request_template.md:11]: I agree, which is why I have [originally proposed a template](https://github.com/Mbed-TLS/mbedtls/pull/9008#issuecomment-2084739951) that had a justifiation after the checkboxes and the linkage. 

Saying why something is not where it should be, can be ingored by tooling but is very helpfull for the gatekeeper verifyin the validity of the template.

File: .github/pull_request_template.md
Comment (minosgalanakis) [reply .github/pull_request_template.md:12]: I would expect this to be filled with `not required` because this is the backport as discussed [above ](https://github.com/Mbed-TLS/mbedtls/pull/9080#discussion_r1595435482).

And again if someone does a provided and self references, that should also work.

But I would refrain from changing the keywords for each entry, since that comes with a significant cost of maintaining the parsing script. 


code like that won't work if each of the lines is unique

~~~~~~~
checkbox_rex = re.complile(r'....')

... find anchors for start/end of checkboxes

for line in text:
  checkbox_rex .search(line)
~~~~~~~

File: .github/pull_request_template.md
Comment (minosgalanakis) [reply .github/pull_request_template.md:15]: To be honest I am a bit torn with Tests and changelog. Yes you are right with the idea of that they should be self contained, and either present or explicitely stated why they are not.

And unlike the linking with backports being on the PR template does not offer much more information than if we used two simple labels such as `no-changelog`, `no-tests` to indicate the use-cases that are not required.

I am happy to go whichever direction the team wants.

File: .github/pull_request_template.md
Comment (yanesca) [reply .github/pull_request_template.md:11]: I agree that we should indicate somehow that we require justification here. I like `~~not required because~~` better than the original `~~not required~~ (comments)` because it is more suggestive of the purpose. Maybe adding a semicolon (`~~not required because:~~`) would make it more likely that whoever edits them keeps the keywords.

File: .github/pull_request_template.md
Comment (yanesca) [reply .github/pull_request_template.md:12]: How about something like this: `provided: #PR, ~not required because:~`

File: .github/pull_request_template.md
Comment (yanesca) [reply .github/pull_request_template.md:15]: In an ideal world, we would always add complete testing with bug fixes and new features. And I agree that we shouldn't encourage adding less testing. However, in practice we have exceptions more often than we would like to: bug fixes where adding tests is prohibitively expensive, new feature development where only limited testing can be added at that point (eg. unit tests or compatibility tests, simply because we don't have enough of the feature implemented to do more, or for practical reasons where we can't keep the PRs small and reviewable in any other way than only adding smoke tests and extensive testing comes in a separate PR) I thought that this line could be useful for providing justification in these cases. As a reviewer or a gatekeeper I would prefer to see some justification as opposed to an empty ~tests` section in the template.

**Review by yanesca — APPROVED** [review]
LGTM

File: .github/pull_request_template.md
Comment (gilles-peskine-arm) [inline .github/pull_request_template.md:11]: Please add a space at the end of the line so we can start writing the part after “because” straight away. Applies to all the lines with “because”.

File: .github/pull_request_template.md
Comment (gilles-peskine-arm) [inline .github/pull_request_template.md:12]: Please don't abbreviate the branch name.
```suggestion
- [ ] **development PR** provided #, ~not required because:~
```

File: .github/pull_request_template.md
Comment (gilles-peskine-arm) [inline .github/pull_request_template.md:13]: A common case will be that a backport is required, but not initially done. Normally we only start working on the backport after the development PR is approved. So maybe initially it should be

```suggestion
- [ ] **3.6 PR** TODO | not required because 
```

File: .github/pull_request_template.md
Comment (gilles-peskine-arm) [inline .github/pull_request_template.md:12]: In a pull request to `development`, it's not clear what we expect people to do. Remove the line? Leave the template alone? Edit to “this is it”? We should give some concise but clear guidance. (Not easy, I know.)

Maybe this?

```suggestion
- [ ] **development PR** here | TODO | provided # | not required because 
```

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
More suggestions to try and make this something that people actually follow.

File: .github/pull_request_template.md
Comment (minosgalanakis) [reply .github/pull_request_template.md:13]: I am againist that since it adds complexity both to the user facing instuctions, as well to the parsing scripts. When something is in TODO  state we can keep the current notation, that the checkbox is unchecked and unmodified.

File: .github/pull_request_template.md
Comment (minosgalanakis) [reply .github/pull_request_template.md:12]: The same as with tests. Use provided without a PR link. I will add a segment in the description if needed.

File: .github/pull_request_template.md
Comment (yanesca) [inline .github/pull_request_template.md:12]: What does that `,` at the end of `provided,` mean? Are we expecting some more information there if that branch is  taken?

File: .github/pull_request_template.md
Comment (yanesca) [inline .github/pull_request_template.md:13]: At other lines we have a `#` after the word `provided` to signal that we expect a PR number. Why don't we need that here?

File: .github/pull_request_template.md
Comment (yanesca) [inline .github/pull_request_template.md:10]: To ease confusion we might want to give instruction on what to do with the dangling # symbol if a PR reference is not needed. Remove it or leave it? I don't mind either way, I just think we should make it clear:

```suggestion
Use # to link to relevant PRs if the provided content is not part the present PR. If the provided content is part of the present PR remove the # symbol.
```

**Review by yanesca — CHANGES_REQUESTED** [review]
I left a couple of questions and remarks.

File: .github/pull_request_template.md
Comment (minosgalanakis) [reply .github/pull_request_template.md:13]: The idea is:

If the pr is a development pr
- [x] **development PR** provided | not required because: 

If the pr is a backport pr or a framework which needs to link to a dev pr
- [x] **development PR** provided # | not required because: 

But I agree it should be consistent so I will add the #

File: .github/pull_request_template.md
Comment (minosgalanakis) [reply .github/pull_request_template.md]: That was a accidental and has been removed.

**Review by yanesca — DISMISSED** [review]
LGTM

File: .github/pull_request_template.md
Comment (gilles-peskine-arm) [inline .github/pull_request_template.md:14]: The framework link will be to another repo. We don't want people to just copy the number: that would create a link in this repo.
```suggestion
- [ ] **framework PR** provided Mbed-TLS/mbedtls-framework# | not required because: 
```

Comment (gilles-peskine-arm) [inline .github/pull_request_template.md:14]: We don't need a reason for framework not provided.
```suggestion
- [ ] **framework PR** provided # | not required
```

File: .github/pull_request_template.md
Comment (gilles-peskine-arm) [inline .github/pull_request_template.md:17]: Covered elsewhere is exceptional for tests and we shouldn't suggest that it's ok.
```suggestion
- **tests**  provided | not required because: 
```

Comment (gilles-peskine-arm) [inline .github/pull_request_template.md:17]: Sometimes there is testing that can't be automated. As a reviewer and gatekeeper, I want to see that in a standard place in the pull request description, so that there's no risk of missing it when it's buried in a long description.
```suggestion
- **tests**  provided | not required because: | additional check that can't be automated: 
```

**Review by gilles-peskine-arm — CHANGES_REQUESTED** [review]
The different items have different ways of being fulfilled. We shouldn't try to impose uniformity where the situations are different.

File: .github/pull_request_template.md
Comment (minosgalanakis) [reply .github/pull_request_template.md:17]: This was discussed in the design review and the reference PR before. In theory it is exceptional but in practise it is way more common, especially by contributions raised by us, where a lot of complicated things are spread into multiple prs. 

I will remove it so we can move forward, but I expect to see the justification of covered in the because section.

e.g
    - **tests**  not required because:  they are covered at #

Comment (minosgalanakis) [reply .github/pull_request_template.md:17]: As above I will not be adding this section, since it can be covered in the justification section.

- **tests**  not required because:  additional check that can't be automated.

**Review by gilles-peskine-arm — APPROVED** [review]
LGTM (to be potentially revised in a few months after it's field-tested)

File: .github/pull_request_template.md
Comment (gilles-peskine-arm) [reply .github/pull_request_template.md]: There are sometimes additional checks but also some tests are provided. But ok, that's not a blocker, let's see how it goes.

**Review by tom-daubney-arm — APPROVED** [review]
LGTM. I think the instructions are clear.

File: .github/pull_request_template.md
Comment (tom-cosgrove-arm) [inline .github/pull_request_template.md:17]: How come there is no checkbox by `tests`?

Comment (gilles-peskine-arm) [reply .github/pull_request_template.md:17]: There's a check box next to PR links because often that PR won't be present from the start. There's a check box next to changelog because we often forget about it so we want an affirmative step about the changelog. Neither consideration applies to tests.

**Comment by gilles-peskine-arm** [comment]
> the fact the checking the boxes should be done by the gatekeeper, not the PR author or reviewers - but the PR author should provide justifications that reviewers should review.

In practice, I expect that submitters won't always pay attention to whether they're supposed to tick the boxes or not. There's no access control or time limit on toggling the boxes and I've even done it by accident when using the mouse to click to focus at least once. So it's up to the gatekeeper to check that everything is fulfilled when merging. And release preparation will need to double check that the ticks are correct.

**Pull request merged**
